src/main/java/spoon/support/reflect/declaration/CtPackageImpl.java

```diff
@@ -45,13 +45,7 @@ protected CtRole getRole() {
 
 		@Override
 		public CtPackage put(String simpleName, CtPackage pack) {
-			if (pack == null) {
-				return null;
-			}
-			// they are the same
-			if (CtPackageImpl.this.getQualifiedName().equals(pack.getQualifiedName())) {
-				addAllTypes(pack, CtPackageImpl.this);
-				addAllPackages(pack, CtPackageImpl.this);
```

Comment:
The comment implies that the package A to which we are trying to add the package B are the same by comparing their qualified names. This simply does not make sense, as the act of add B as a subpackage to A _changes_ its qualified name to `A.B`.

It's simply impossible for `A` and `B` to end up with the same qualified name after adding `B` to `A` as the qualified name of `A` will always be prefixed to `B`.

Reply:
I was quite confused about the check as well when adjusting this class. It sounded like this was meant to merge equal packages. If I didn't miss anything, the two names *can* be the same at the time of the check, if you have two parallel package hierarchies. The linking is only done in the `super` call, so the parent isn't yet replaced.

I think this could happen with the module system (though the overlap is purely due to spoon's creation of synthetic packages). 

I was afraid to touch it back then, but it doesn't seem particularly likely that code depends on this.

Reply:
> I was quite confused about the check as well when adjusting this class. It sounded like this was meant to merge equal packages. If I didn't miss anything, the two names _can_ be the same at the time of the check, if you have two parallel package hierarchies.

Absolutely, the names can be the same (e.g. `A = B`, which is precisely the bug reported in #4764). It doesn't matter, however, because the semantics of this method is to add `B` as a subpackage of `A`. A subpackage of `A` cannot have the same qualified name as `A`, as `A`'s qualified name is prepended to all of its supbackages. There simply cannot be a name collision; the action that this method performs makes that impossible.

It just.. doesn't make sense.

Reply:
What I was trying to convey is, that you can hit this code path in existing code (e.g.):
```java
		Launcher launcher = new Launcher();
		Factory factory = launcher.getFactory();
		CtModule mod = factory.Module().getOrCreate("second");
		CtPackage pack1 = factory.Package().getOrCreate("test", factory.Module().getUnnamedModule());
		CtPackage pack2 = factory.Package().getOrCreate("test", mod);

		pack1.addPackage(pack2);
```
or if you obtain a second package another way. The existing code would merge this, your code would nest them. This is an observable semantic difference. I just wanted to point out, that this *can happen* and is *not* impossible and why I left that code in there when refactoring the ModelSet.

I do agree though, that the new semantics are likely closer to what the user intended and probably won't do any harm. You just *might* break something, as you change observable behavior.

Reply:
For clarity, if you run `Launcher().getFactory().Package().getOrCreate("A.A")`, you get back a package with the qualified name `A` (before removal of the discussed code path).

Reply:
Oh no need to apologize, I appreciate your attention to detail :)